subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt.c

```diff
@@ -354,44 +398,186 @@ static int fs_mgmt_file_upload(struct smp_streamer *ctxt)
 		if (len == ULLONG_MAX) {
 			return MGMT_ERR_EINVAL;
 		}
+	}
 
-		fs_mgmt_ctxt.uploading = true;
-		fs_mgmt_ctxt.off = 0;
-		fs_mgmt_ctxt.len = len;
-	} else {
-		if (!fs_mgmt_ctxt.uploading) {
-			return MGMT_ERR_EINVAL;
+	if (k_sem_take(&fs_mgmt_ctxt.lock_sem, FILE_SEMAPHORE_MAX_TAKE_TIME)) {
+		return MGMT_ERR_EBUSY;
+	}
+
+	/* Check if there is an active upload in progress */
+	if (ctxt->smpt != fs_mgmt_ctxt.transport) {
+		/* Different transport, clean up */
+		close_file = true;
+		open_file = true;
+	} else if (fs_mgmt_ctxt.state == STATE_UPLOAD) {
+		/* Check if this is the same file */
+		if (strcmp(file_name, fs_mgmt_ctxt.path)) {
+			/* Different file, we need to clean up */
+			close_file = true;
+			open_file = true;
 		}
+	} else if (fs_mgmt_ctxt.state == STATE_DOWNLOAD) {
+		/* Different mode, we need to clean up */
+		close_file = true;
+		open_file = true;
+	} else {
+		open_file = true;
+	}
+
+	/* Close existing file if it was open */
+	if (close_file) {
+		fs_mgmt_cleanup();
+	}
+
+	/* Open new file */
+	if (open_file) {
+		fs_mgmt_ctxt.off = 0;
+		fs_file_t_init(&fs_mgmt_ctxt.file);
+		rc = fs_open(&fs_mgmt_ctxt.file, file_name, FS_O_CREATE | FS_O_WRITE);
 
-		if (off != fs_mgmt_ctxt.off) {
-			/* Invalid offset.  Drop the data and send the expected offset. */
-			return fs_mgmt_file_rsp(zse, MGMT_ERR_EINVAL, fs_mgmt_ctxt.off);
+		if (rc != 0) {
+			k_sem_give(&fs_mgmt_ctxt.lock_sem);
+			return MGMT_ERR_ENOENT;
 		}
+
+		strcpy(fs_mgmt_ctxt.path, file_name);
+		fs_mgmt_ctxt.state = STATE_UPLOAD;
+		fs_mgmt_ctxt.transport = ctxt->smpt;
+	}
+
+	if (off == 0) {
+		fs_mgmt_ctxt.len = len;
 	}
 
-	new_off = fs_mgmt_ctxt.off + file_data.len;
+	new_off = off + file_data.len;
+
 	if (new_off > fs_mgmt_ctxt.len) {
-		/* Data exceeds image length. */
-		return MGMT_ERR_EINVAL;
+		/* Data exceeds image length, but check if this is because we need to re-create
+		 * the variable.
+		 */
+		if (fs_mgmt_ctxt.len == 0) {
```

Comment:
I think that here the server should respond with the value read by `fs_mgmt_file.len`  and wait for next client frame with expected offset.
This would follow the previous logic where in case when offset differs the from expected, the code will request the one it expects, with the difference that next expected offset is taken from context or file len, if that in context is not available.

Reply:
Is it even worth keeping last offset within context? 
This makes upload behave differently when file is held open through entire upload and when idle close happens.
Is the `len` field of request with `off == 0` request still relevant since we close the file when system gets idle?
In case when we can hold file open this will upload till the `len` is achieved, then file gets closed. But you can still send next request with `off` == len of file, and extend file further.
Maybe with this pr, we should obsolete the `len` parameter , as a file length and change behaviour to:
 - if `off < 0` and `off < file_len_in_file_system` we truncate file and start write from beginning;
 - if `off != 0` and `off > file_len_in_file_system` we return expected offset, which is file len;
 - if `off == file_len_in_file_system` we add chunk to the end of file;
This will significantly simplify logic and should not break mcumgr-cli, as it is expecting offset of last successful write, which we can always provide.

Reply:
Since the client should be keeping track of the offset and changing it only depending upon what response it gets from the server, it's probably safe to not use it

Reply:
Reworked and fixed